### PR TITLE
Formatting Feature - First Pass

### PR DESCRIPTION
**Pull request by tarik-brown**
This pull request contains all of the plumbing needed to access the vcFormat settings

### CODE REVIEW

File: Extension/.vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+    "configurations": [
```
Comment (bobbrow) [inline Extension/.vscode/c_cpp_properties.json]: I don't think this file should be checked in

File: Extension/package.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "name": "cpptools",
   "displayName": "C/C++",
   "description": "C/C++ IntelliSense, debugging, and code browsing.",
-  "version": "0.29.0-master",
+  "version": "0.28.2-master",
```
Comment (bobbrow) [inline Extension/package.json]: revert this (or update it to 0.30.0-master)

File: Extension/package.json [hunk 3]
```diff
@@ -39,7 +39,31 @@
     "Snippets"
   ],
   "activationEvents": [
-    "*"
```
Comment (bobbrow) [inline Extension/package.json]: revert the "activationEvents" back to "*"

File: Extension/package.json [hunk 4]
```diff
@@ -93,6 +117,454 @@
           "description": "%c_cpp.configuration.clang_format_style.description%",
           "scope": "machine-overridable"
         },
+        "C_Cpp.test_root": {
```
Comment (bobbrow) [inline Extension/package.json]: `test_root` should be removed

File: Extension/package.json [hunk 5]
```diff
@@ -93,6 +117,454 @@
...
+              "type": "string",
+              "enum": [
+                "Enabled",
+                "Disabled"
+              ],
+              "default": "Enabled"
+            }
+          }
+        },
+        "C_Cpp.formatting.engine": {
```
Comment (bobbrow) [inline Extension/package.json]: I thought we were going to keep C_Cpp.formatting a bit longer and you would just add the new enum values to it.

File: Extension/yarn-error.log [hunk 6]
```diff
@@ -0,0 +1,8254 @@
+Arguments:
```
Comment (bobbrow) [inline Extension/yarn-error.log]: This should be removed (or added to .gitignore)

File: Extension/package.nls.json [hunk 7]
```diff
@@ -17,6 +17,64 @@
     "c_cpp.command.rescanWorkspace.title": "Rescan Workspace",
     "c_cpp.command.vcpkgClipboardInstallSuggested.title": "Copy vcpkg install command to clipboard",
     "c_cpp.command.vcpkgOnlineHelpSuggested.title": "Visit the vcpkg help page",
+    "c_cpp.configuration.formatting.engine.description": "Select Formatting Engine",
```
Comment (Colengms) [inline Extension/package.nls.json]: I think these should be more descriptive.  i.e. "Specifies which engine is used to perform formatting".  "Indicates whether ...", etc.

File: Extension/src/LanguageServer/client.ts [hunk 8]
```diff
@@ -1133,6 +1133,7 @@ export class DefaultClient implements Client {
         const settings_filesExclude: (vscode.WorkspaceConfiguration | undefined)[] = [];
         const settings_searchExclude: (vscode.WorkspaceConfiguration | undefined)[] = [];
         const settings_editorTabSize: (number | undefined)[] = [];
+      //  const settings_editorSize: (number | undefined)[] = [];
```
Comment (Colengms) [inline Extension/src/LanguageServer/client.ts]: Remove commented out code before checkin.

File: Extension/package.json [hunk 9]
```diff
@@ -1588,6 +2050,7 @@
     "@typescript-eslint/parser": "^2.19.2",
     "async-child-process": "^1.1.1",
     "await-notify": "^1.0.1",
+    "editorconfig": "^0.15.3",
```
Comment (Colengms) [inline Extension/package.json]: Will you be checking code in prior to starting on the editorconfig portion?  If so, I would wait to add this until after you've checked in this first batch of changes.

File: Extension/src/LanguageServer/settings.ts [hunk 10]
```diff
@@ -165,14 +164,249 @@ export class CppSettings extends Settings {
...
+
+    public get vcFormatNewlineBeforeOpenBraceBlock() : string | undefined {
+        return super.Section.get<string>("vcFormat.newLine.beforeOpenBrace.block");
+    }
+
+    public get vcFormatNewlineBeforeOpenBraceLambda() : string | undefined {
+        return super.Section.get<string>("vcFormat.newLine.beforeOpenBrace.lamda");
+    }
+
+    public get vcFormatNewlineScopeBracesOnSeparateLines() : boolean | undefined {
```
Comment (Colengms) [inline Extension/src/LanguageServer/settings.ts]: Using `=== true` removes the potential for an `undefined` result.  You'll want to remove ` | undefined` from the return type of these functions that use `=== true`.

File: Extension/src/LanguageServer/settings.ts [hunk 11]
```diff
@@ -203,6 +437,7 @@ export class OtherSettings {
     }
 
     public get editorTabSize(): number | undefined { return vscode.workspace.getConfiguration("editor", this.resource).get<number>("tabSize"); }
```
Comment (Colengms) [inline Extension/src/LanguageServer/settings.ts]: I believe editorTabSize and editorInsertSpaces, and everything related to them, can be removed, as they are not used.  The right values to use for these are instead passed with every formatting request.

File: Extension/yarn.lock [hunk 12]
```diff
@@ -1235,7 +1235,7 @@ commander@2.11.0:
   resolved "https://registry.yarnpkg.com/commander/-/commander-2.11.0.tgz#157152fd1e7a6c8d98a5b715cf376df928004563"
```
Comment (Colengms) [inline Extension/yarn.lock]: If you remove the editorconfig module, you'll also want to revert yarn.lock.

**Review by Colengms — CHANGES_REQUESTED** [review]

File: Extension/package.json (hunk 3 above)
Comment (sean-mcmanus) [reply Extension/package.json]: Looks like an extra space is still there. Might be worth running a format document on the package.json file.

File: Extension/package.json [hunk 13]
```diff
@@ -93,6 +93,430 @@
           "description": "%c_cpp.configuration.clang_format_style.description%",
           "scope": "machine-overridable"
         },
+        "C_Cpp.formatting": {
+          "type": "string",
+          "enum": [
+            "clangFormat",
```
Comment (sean-mcmanus) [inline Extension/package.json:99]: The "Default" enum was removed -- is that breaking change going to be an issue? What happens for existing users that have "Default" set?

File: Extension/package.json [hunk 14]
```diff
@@ -1966,4 +2381,4 @@
       "binaries": []
     }
   ]
-}
```
Comment (sean-mcmanus) [inline Extension/package.json]: Should there be a newline or not? I seem to recall it going back and forth.

File: Extension/package.nls.json [hunk 15]
```diff
@@ -17,6 +17,64 @@
     "c_cpp.command.rescanWorkspace.title": "Rescan Workspace",
     "c_cpp.command.vcpkgClipboardInstallSuggested.title": "Copy vcpkg install command to clipboard",
     "c_cpp.command.vcpkgOnlineHelpSuggested.title": "Visit the vcpkg help page",
+    "c_cpp.configuration.formatting.description": "Select Formatting Engine",
+    "c_cpp.configuration.vcFormat.indent.braces.description": "Indent Braces",
+    "c_cpp.configuration.vcFormat.multiLineRelativeTo.description" : "Indent MultiLine",
```
Comment (sean-mcmanus) [inline Extension/package.nls.json]: Can you run a format document on this file? I see ": " and " : "?

File: Extension/src/LanguageServer/client.ts [hunk 16]
```diff
@@ -2013,7 +2212,7 @@ export class DefaultClient implements Client {
                         }
                     }
                 }
-            });
+            });5
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: 5?

File: Extension/src/LanguageServer/client.ts [hunk 17]
```diff
@@ -1375,6 +1562,18 @@ export class DefaultClient implements Client {
         const settings: any = {
             C_Cpp: {
                 ...cppSettingsScoped,
+                vcFormat: {
+                  //  ...vscode.workspace.getConfiguration("C_Cpp.vcFormat",this.RootUri),
+                    indent: vscode.workspace.getConfiguration("C_Cpp.vcFormat.indent",this.RootUri),
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/client.ts]: If you run the "lint" npm script it should give an error about the missing space after the ",".

File: Extension/src/LanguageServer/settings.ts [hunk 18]
```diff
@@ -30,23 +30,23 @@ class Settings {
     protected get Section(): vscode.WorkspaceConfiguration { return this.settings; }
 
     protected getWithFallback<T>(section: string, deprecatedSection: string): T {
-        const info: any = this.settings.inspect<T>(section);
+        let info: any = this.settings.inspect<T>(section);
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/settings.ts]: I think this should still be "const". The linter should complain.

File: Extension/src/LanguageServer/client.ts (hunk 17 above)
Comment (bobbrow) [reply Extension/src/LanguageServer/client.ts]: please also delete the commented out code above this line.

File: Extension/package.nls.json [hunk 19]
```diff
@@ -17,6 +17,64 @@
     "c_cpp.command.rescanWorkspace.title": "Rescan Workspace",
     "c_cpp.command.vcpkgClipboardInstallSuggested.title": "Copy vcpkg install command to clipboard",
     "c_cpp.command.vcpkgOnlineHelpSuggested.title": "Visit the vcpkg help page",
+    "c_cpp.configuration.formatting.description": "Select Formatting Engine",
```
Comment (Colengms) [inline Extension/package.nls.json]: I think you'll want to specify more here.  There was an existing description for this that was removed, that would be a good starting point.

File: Extension/package.nls.json [hunk 20]
```diff
@@ -78,8 +135,8 @@
     "c_cpp.debuggers.pipeTransport.pipeCwd.description": "The fully qualified path to the working directory for the pipe program.",
     "c_cpp.debuggers.pipeTransport.pipeProgram.description": "The fully qualified pipe command to execute.",
     "c_cpp.debuggers.pipeTransport.pipeArgs.description": "Command line arguments passed to the pipe program to configure the connection.",
-    "c_cpp.debuggers.pipeTransport.pipeEnv.description": "Environment variables passed to the pipe program.",
-    "c_cpp.debuggers.pipeTransport.quoteArgs.description": "If the pipeProgram's individual arguments contain characters (such as spaces or tabs), should it be quoted? If 'false', the debugger command will no longer be automatically quoted. \nDefault is 'true'.",
+    "c_cpp.debuggers.pipeTransport.pipeEnv.description": "Environment variables passed to the pipe program.",   
+     "c_cpp.debuggers.pipeTransport.quoteArgs.description": "If the pipeProgram's individual arguments contain characters (such as spaces or tabs), should it be quoted? If 'false', the debugger command will no longer be automatically quoted. \nDefault is 'true'.",
```
Comment (Colengms) [inline Extension/package.nls.json]: It looks like some stray spaces were added here.

**Review by Colengms — CHANGES_REQUESTED** [review]

File: Extension/package.json (hunk 13 above)
Comment (bobbrow) [reply Extension/package.json:99]: Default was supposed to stay. It has been added back in the latest push.

**Review by Colengms — APPROVED** [review]

**Pull request merged**